tcp/tcp.c

```diff
@@ -0,0 +1,382 @@
+			espconn_disconnect(client->pCon);
+		}
+
+		cmdResponseStart(CMD_RESP_CB, client->resp_cb, 3);
+		cmdResponseBody(&cb_type, 1);	
+		cmdResponseBody(&clientNum, 1);
+		cmdResponseBody(&sentDataLen, 2);
+		cmdResponseEnd();
+	}
+}
+
+// Connection is disconnected
+static void ICACHE_FLASH_ATTR
+tcpclient_discon_cb(void *arg) {
+	struct espconn *pespconn = (struct espconn *)arg;
+	TcpClient* client = (TcpClient *)pespconn->reverse;
+
+	uint8_t clientNum = client->conn_num;
+	uint8_t cb_type = 3;
+	sint16 _status = 0;
+	DBG_TCP("TCP #%d: Disconnect\n", client-tcpClient);
+	// free the data buffer, if we have one
+	if (client->data) os_free(client->data);
+	client->data = 0;
+	cmdResponseStart(CMD_RESP_CB, client->resp_cb, 3);
+	cmdResponseBody(&cb_type, 1);	
+	cmdResponseBody(&clientNum, 1);
+	cmdResponseBody(&_status, 2);
+	cmdResponseEnd();
+}
+
+// Connection was reset
+static void ICACHE_FLASH_ATTR
+tcpclient_recon_cb(void *arg, sint8 errType) {
+	struct espconn *pCon = (struct espconn *)arg;
+	TcpClient* client = (TcpClient *)pCon->reverse;
+
+	uint8_t clientNum = client->conn_num;
+	uint8_t cb_type = 2;
+	sint16 _errType = errType;
+	os_printf("TCP #%d: conn reset, err=%d\n", client-tcpClient, _errType);
+	cmdResponseStart(CMD_RESP_CB, client->resp_cb, 3);
+	cmdResponseBody(&cb_type, 1);	
+	cmdResponseBody(&clientNum, 1);
+	cmdResponseBody(&_errType, 2);
+	cmdResponseEnd();
+	// free the data buffer, if we have one
+	if (client->data) os_free(client->data);
+	client->data = 0;
+}
+
+// Connection is done
+static void ICACHE_FLASH_ATTR
+tcpclient_connect_cb(void *arg) {
+	struct espconn *pCon = (struct espconn *)arg;
+	TcpClient* client = (TcpClient *)pCon->reverse;
+
+	uint8_t clientNum = client->conn_num;
+	uint8_t cb_type = 3;
+	sint16 _status = 1;
+	DBG_TCP("TCP #%d: connected socket mode = %d\n", client-tcpClient, client->sock_mode);
+	espconn_regist_disconcb(client->pCon, tcpclient_discon_cb);
+	espconn_regist_recvcb(client->pCon, tcpclient_recv_cb);
+	espconn_regist_sentcb(client->pCon, tcpclient_sent_cb);
+
+	DBG_TCP("TCP #%d: sending %d\n", client-tcpClient, client->data_sent);
+	if (client->sock_mode != SOCKET_SERVER) { // Send data after established connection only in client mode
+		client->data_sent = client->data_len <= 1400 ? client->data_len : 1400;
+		DBG_TCP("TCP #%d: sending %d\n", client-tcpClient, client->data_sent);
+		espconn_send(client->pCon, (uint8_t*)client->data, client->data_sent);
+	}
+
+	cmdResponseStart(CMD_RESP_CB, client->resp_cb, 3);
+	cmdResponseBody(&cb_type, 1);	
+	cmdResponseBody(&clientNum, 1);
+	cmdResponseBody(&_status, 2);
+	cmdResponseEnd();
+}
+
+static void ICACHE_FLASH_ATTR
+tcp_dns_found(const char *name, ip_addr_t *ipaddr, void *arg) {
+	struct espconn *pConn = (struct espconn *)arg;
+	TcpClient* client = (TcpClient *)pConn->reverse;
+
+	if(ipaddr == NULL) {
+		os_printf("TCP #%d DNS: Got no ip, try to reconnect\n", client-tcpClient);
+		return;
```

Comment:
how does the attached uC find out about this, seems like the error is silently being swallowed?